src/ripple/basics/impl/Log.cpp

```diff
@@ -289,23 +289,41 @@ Logs::fromString (std::string const& s)
     return lsINVALID;
 }
 
-// Replace the first secret, if any, with asterisks
+// Attempt to prevent sensitive information from appearing in log files by
+// redacting it with asterisks.
 std::string
 Logs::scrub (std::string s)
 {
-    using namespace std;
-    char const* secretToken = "\"secret\"";
-    // Look for the first occurrence of "secret" in the string.
-    size_t startingPosition = s.find (secretToken);
-    if (startingPosition != string::npos)
+    auto scrubber = [&s](char const* token)
     {
-        // Found it, advance past the token.
-        startingPosition += strlen (secretToken);
-        // Replace the next 35 characters at most, without overwriting the end.
-        size_t endingPosition = std::min (startingPosition + 35, s.size () - 1);
-        for (size_t i = startingPosition; i < endingPosition; ++i)
-            s [i] = '*';
-    }
+        auto first = s.find(token);
+
+        // If we have found the specified token, then attempt to isolate the
+        // sensitive data (it's enclosed by double quotes, followed by a comma)
+        // and mask it:
+
+        if (first != std::string::npos)
+            first = s.find ('\"', std::min(first + std::strlen(token), s.size() - 1));
+
+        if (first != std::string::npos)
+        {
+            auto last = s.find(',', first);
+
+            if (last == std::string::npos)
+                last = s.size();
+
+            s.replace (first, last - first, last - first, '*');
+        }
+    };
```

Comment:
On the `find('\"',...)` there is no need to limit the starting search position to the size of the string.  The `find` implementation does this check already.

I recommend searching for the second `"` instead of the `,` after the first `"` because the sensitive value may be the last key/value pair in that level of the jason.

Here is a minor tweak which takes these suggestions into account:
```c++
    auto scrubber = [&s](char const* token)
    {
        auto first = s.find(token);

        // If we have found the specified token, then attempt to isolate the
        // sensitive data (it's enclosed by double quotes)
        // and mask it:

        if (first != std::string::npos)
        {
            first = s.find ('\"', first + std::strlen(token));
            if (first != std::string::npos)
            {
                auto last = s.find('\"', first+1);

                if (last == std::string::npos)
                    last = s.size();
                else
                    ++last;

                s.replace (first, last - first, last - first, '*');
            }
        }
    };
```

Reply:
Unlikely to happen but worth mentioning that subsequent occurrences of the token to not get masked.

Reply:
Since this is called for every write to the log, as as written it does seven passes of the string, how about using a regex to do one pass. Something like:
```
std::string
scrub (std::string s)
{
    static std::regex const r{
        R"regex("(seed|seed_hex|secret|master_key|master_seed|master_seed_hex|passphrase)")regex",
        std::regex_constants::optimize};

    std::smatch m;
    std::smatch::difference_type p = 0;
    while (1)
    {
        if (!std::regex_search(s.cbegin() + p, s.cend(), m, r))
            return s;
        p = m.position() + p;
        p = s.find('\"', p + m.length());
        if (p == std::string::npos)
            return s;

        auto last = s.find(',', p);
        if (last == std::string::npos)
            last = s.size();
        s.replace (p, last - p, last - p, '*');
        p = last;
    }

    return s;
}

```

Reply:
Data point:

I took my implementation and tested it against the input string in the bug report that is prefixed by "2018-Jan-25 15:02:38 Server:DBG Reply:".  Then I took Scott's implementation, changed it to search for the second `"` instead of `,`, and tested it against the same string.  Compiled with clang/libc++ at `-O3`.  Mine took about 50us and Scott's about 230us.  Other regex implementations might be much faster.  I wrote the libc++ regex, and I know it sucks. :-\

Reply:
@HowardHinnant Does that include the time to compile the regex?

Reply:
(Scott and I are working offline on this)

Reply:
Yeah, the single-pass regex solution is much slower than the multi-pass hand coded solution. I'm actually surprised. Props to @HowardHinnant for actually timing this. Here's the result I got on gcc for 10000 iterations in release mode:
```
No regex: 16789424
Regex:    98181946
```

Reply:
@HowardHinnant I believe your code masks the first double quote of the secret. 
Should be
```
        if (first != std::string::npos)
        {
            first = s.find ('\"', first + std::strlen(token));
            if (first != std::string::npos)
            {
                auto last = s.find('\"', ++first);
                if (last == std::string::npos)
                    last = s.size();
                s.replace (first, last - first, last - first, '*');
            }
        }
```

Reply:
Right.  It masks the beginning _and_ ending quote.  I did that because that's what the original code did.  If we want `"****"` instead, sounds good to me.

Reply:
@HowardHinnant Ah I see. I incorrectly assumed we only wanted to mask within the quotes to keep the json valid.

Reply:
Personally I think `"***"` would look better, but I'm not going to loose any sleep either way. :-)

Reply:
I agree. It would be, theoretically, marginally more secure (you don't leak the length). However, that would necessitate a lot more copying. I'm happy to tweak that if others think it's important.

Reply:
I think masking only within the quotes is a good idea. The tweak I previously mentioned to @HowardHinnant may be all we need.

Reply:
I wrote another version...
```
    auto scrubber = [&s](char const* prefix,
        std::vector<char const*> const& endings)
    {
        auto first = s.find(prefix);
        if (first == std::string::npos)
            return;
        size_t const offset = first + std::strlen(prefix);
        for (auto const& e : endings)
        {
            auto const len = std::strlen(e);
            if ((s.compare(offset, len, e)) == 0)
            {
                first = s.find('\"', offset + len);
                if (first != std::string::npos)
                {
                    auto last = s.find('\"', first + 1);
                    if (last == std::string::npos)
                        last = s.size();
                    else
                        ++last;
                    s.replace(first, last - first, last - first, '*');
                }
                return;
            }
        }
    };
```

called as such...
```
        scrubber("\"master_", { "key\"", "seed\"", "seed_hex\"" });
        scrubber("\"passphrase", { "\"" });
```

However it is about 30% slower than @HowardHinnant's version.